feature: openpmd writer for 3d hockney solver

s-sajid-ali opened the pull request:
Draft PR to run CI during development.

s-sajid-ali committed:
OpenPMD write particles

	modified:   CMakeLists.txt
	modified:   src/synergia/collective/CMakeLists.txt
	new file:   src/synergia/collective/openpmd_writer.cc
	new file:   src/synergia/collective/openpmd_writer.h
	modified:   src/synergia/collective/rectangular_grid_domain.h

s-sajid-ali committed:
Use openpmd in sc3d-open-hockney

s-sajid-ali committed:
minor

	modified:   src/synergia/bunch/bunch.cc
	modified:   src/synergia/bunch/bunch.h
	modified:   src/synergia/bunch/bunch_particles.cc
	modified:   src/synergia/bunch/bunch_particles.h
	modified:   src/synergia/bunch/bunch_pywrap.cc
	modified:   src/synergia/bunch/bunch_train.cc
	modified:   src/synergia/bunch/bunch_train.h
	modified:   src/synergia/bunch/core_diagnostics.cc
	modified:   src/synergia/bunch/core_diagnostics.h
	modified:   src/synergia/bunch/diagnostics.h
	modified:   src/synergia/bunch/diagnostics_basic.cc
	modified:   src/synergia/bunch/diagnostics_basic.h
	modified:   src/synergia/bunch/diagnostics_bulk_track.cc
	modified:   src/synergia/bunch/diagnostics_bulk_track.h
	modified:   src/synergia/bunch/diagnostics_file.cc
	modified:   src/synergia/bunch/diagnostics_file.h
	modified:   src/synergia/bunch/diagnostics_full2.cc
	modified:   src/synergia/bunch/diagnostics_full2.h
	modified:   src/synergia/bunch/diagnostics_full2_host.cc
	modified:   src/synergia/bunch/diagnostics_loss.cc
	modified:   src/synergia/bunch/diagnostics_loss.h
	modified:   src/synergia/bunch/diagnostics_particles.cc
	modified:   src/synergia/bunch/diagnostics_particles.h
	modified:   src/synergia/bunch/diagnostics_py.cc
	modified:   src/synergia/bunch/diagnostics_py.h
	modified:   src/synergia/bunch/diagnostics_track.cc
	modified:   src/synergia/bunch/diagnostics_track.h
	modified:   src/synergia/bunch/diagnostics_worker.cc
	modified:   src/synergia/bunch/diagnostics_worker.h
	modified:   src/synergia/bunch/fixed_t_z_converter.cc
	modified:   src/synergia/bunch/fixed_t_z_converter.h
	modified:   src/synergia/bunch/period.cc
	modified:   src/synergia/bunch/period.h
	modified:   src/synergia/bunch/populate.h
	modified:   src/synergia/bunch/populate_global.cc
	modified:   src/synergia/bunch/populate_global.h
	modified:   src/synergia/bunch/populate_host.cc
	modified:   src/synergia/collective/openpmd_writer.cc

s-sajid-ali committed:
minor

s-sajid-ali committed:
Do not install OpenPMD fetched via CMake

s-sajid-ali committed:
Do not install OpenPMD fetched via CMake

ax3l commented on CMakeLists.txt:
```diff
@@ -131,6 +131,25 @@ option(
+  set(openPMD_BUILD_CLI_TOOLS OFF)
+  set(openPMD_BUILD_EXAMPLES OFF)
+  set(openPMD_BUILD_TESTING OFF)
+  set(openPMD_INSTALL OFF)
```
For the issue with
```suggestion
  set(openPMD_INSTALL ON)
```
I think the issue we see here might be due to the CMake version in CI, I cannot immediately see the same issue on my Ubuntu with CMake `3.23.2`. What is the CMake version in CI here?
```
CMake Error at /home/runner/work/synergia2/synergia2/build/_deps/openpmd-src/CMakeLists.txt:1080 (install):
  install EXPORT given no DESTINATION!


CMake Error at /home/runner/work/synergia2/synergia2/build/_deps/openpmd-src/CMakeLists.txt:1085 (install):
  install FILES given no DESTINATION!
```
https://github.com/openPMD/openPMD-api/blob/78baa4cc16b04b446787e69059f2b7b27aba136f/CMakeLists.txt#L1080-L1085
I will check more...

ax3l replied:
For the issue with
```suggestion
  set(openPMD_INSTALL ON)
```
I think the issue we see here might be due to the CMake version in CI, I cannot immediately see the same issue on my Ubuntu with CMake `3.23.2`. What is the CMake version in CI here?
```
CMake Error at /home/runner/work/synergia2/synergia2/build/_deps/openpmd-src/CMakeLists.txt:1080 (install):
  install EXPORT given no DESTINATION!


CMake Error at /home/runner/work/synergia2/synergia2/build/_deps/openpmd-src/CMakeLists.txt:1085 (install):
  install FILES given no DESTINATION!
```
https://github.com/openPMD/openPMD-api/blob/78baa4cc16b04b446787e69059f2b7b27aba136f/CMakeLists.txt#L1080-L1085
I will check more...

ax3l replied:
> What is the CMake version in CI here?

CI: CMake 3.22.1 per https://packages.ubuntu.com/jammy/cmake

ax3l replied:
Fix coming via https://github.com/openPMD/openPMD-api/pull/1326

ax3l commented on CMakeLists.txt:
```diff
@@ -131,6 +131,25 @@ option(
+  include(FetchContent)
+  set(CMAKE_POLICY_DEFAULT_CMP0077 NEW)
+  set(openPMD_BUILD_CLI_TOOLS OFF)
+  set(openPMD_BUILD_EXAMPLES OFF)
```
Maybe more usable for the superbuild - and usable with `openPMD_INSTALL=OFF`, is:
```suggestion
  set(openPMD_BUILD_EXAMPLES OFF)
  set(openPMD_BUILD_SHARED_LIBS OFF)
```
that way you don't need to install and find a shared lib for openPMD.

ax3l replied:
Doc update coming here: https://github.com/openPMD/openPMD-api/pull/1325

s-sajid-ali committed:
Update CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

s-sajid-ali committed:
Empty-Commit

s-sajid-ali committed:
bump h5py, install from git

	modified:   .github/workflows/macos-clang.yml
	modified:   .github/workflows/macos-gcc.yml

s-sajid-ali committed:
Empty-Commit to restart CI

s-sajid-ali committed:
CI:macos use latest changes from devel3

s-sajid-ali committed:
Write to OpenPMD from multiple MPI ranks

s-sajid-ali committed:
try fetching HDF5 when hdf5~mpi is found

s-sajid-ali committed:
Congigure OpenPMD after HDF5

s-sajid-ali committed:
Congigure OpenPMD after HDF5

s-sajid-ali committed:
minor

s-sajid-ali commented:
Superseded by #105.
